Add Command::executeCommand()

markstory opened the pull request:
This method enables commands to dispatch each other without needing to use the CommandRunner. This provides commands with the ability to be composed from other commands.

Refs #12549

markstory committed:
Add Command::executeCommand()

This method enables commands to dispatch each other without needing to
use the CommandRunner. This provides commands with the ability to be
composed from other commands.

Refs #12549

ADmad commented on src/Console/Command.php:
```diff
@@ -249,4 +249,34 @@ public function abort($code = self::CODE_ERROR)
+     * @param array $args The arguments to invoke the command with.
+     * @return int|null The exit code or null for success of the command.
+     */
+    public function executeCommand($command, ConsoleIo $io, array $args = [])
```
Can the method name and arguments order be harmonized with `CommandRunner::runCommand()`?
Could the `$io` argument be made optional and internally create a new instance if none provided?

markstory replied:
I could do that. Making a new instance will do bad things in tests which is why I made it a required argument.

markstory replied:
Done now :+1:

markstory committed:
Flip argument order

This puts arguments people use the most first and matches the argument
order of CommandRunner::run()

markstory committed:
Add test for abort case

ADmad commented on src/Console/Command.php:
```diff
@@ -249,4 +249,35 @@ public function abort($code = self::CODE_ERROR)
+     * @param \Cake\Console\ConsoleIo $io The ConsoleIo instance to use for the executed command.
+     * @return int|null The exit code or null for success of the command.
+     */
+    public function executeCommand($command, array $args = [], ?ConsoleIo $io = null)
```
`?ConsoleIo` typehint won't work in PHP < 7.1 :stuck_out_tongue:.

dereuromark replied:
Dropping the ? should be fine here :)

markstory replied:
Derp. I forgot that this was for 3.x

markstory committed:
Remove PHP7 specific syntax

markstory commented:
> Fatal error: Declaration of Cake\TestSuite\TestCase::setUp() must be compatible with PHPUnit\Framework\TestCase::setUp(): void in /home/travis/build/cakephp/cakephp/src/TestSuite/TestCase.php on line 36

Going to rollback the PHPUnit version until we can make a separate effort to do that.
